joblib/test/test_hashing.py

```diff
@@ -408,19 +454,45 @@ def test_hashes_stay_the_same_with_numpy_objects():
         np.arange(100, dtype='<i8').reshape((10, 10))[:, :2],
     ]
 
-    # These expected results have been generated with joblib 0.9.0
-    expected_hashes = [
-        '10a6afc379ca2708acfbaef0ab676eab',
-        '988a7114f337f381393025911ebc823b',
-        'c6809f4b97e35f2fa0ee8d653cbd025c',
-        'b3ad17348e32728a7eb9cda1e7ede438',
-        '927b3e6b0b6a037e8e035bda134e0b05',
-        '108f6ee98e7db19ea2006ffd208f4bf1',
-        'bd48ccaaff28e16e6badee81041b7180'
+    rng = np.random.RandomState(42)
+    # same list of object as previous list. joblib.hash should return the same
+    # hash for to_hash_list_one[i] and to_hash_list_two[i]
+    to_hash_list_two = [
+        rng.randint(-1000, high=1000, size=50).astype('<i8'),
+        tuple(rng.randn(3).astype('<f4') for _ in range(5)),
+        [rng.randn(3).astype('<f4') for _ in range(5)],
+        {
+            -3333: rng.randn(3, 5).astype('<f4'),
+            0: [
+                rng.randint(10, size=20).astype('<i8'),
+                rng.randn(10).astype('<f4')
+            ]
+        },
+        # Non regression cases for https://github.com/joblib/joblib/issues/308.
+        # Generated with joblib 0.9.4.
+        np.arange(100, dtype='<i8').reshape((10, 10)),
+        # Fortran contiguous array
+        np.asfortranarray(np.arange(100, dtype='<i8').reshape((10, 10))),
+        # Non contiguous array
+        np.arange(100, dtype='<i8').reshape((10, 10))[:, :2],
     ]
 
-    for to_hash, expected in zip(to_hash_list, expected_hashes):
-        assert hash(to_hash) == expected
+    e = ProcessPoolExecutor(max_workers=2)
+    for obj_1, obj_2 in zip(to_hash_list_one, to_hash_list_two):
+        # testing consistency of hashing for two copies of the same objects.
+        hash_1, hash_2 = hash(obj_1), hash(obj_2)
+        assert hash_1 == hash_2
+
+        # testing consistency of hashes across python processes
+        hash_1, hash_2 = e.map(hash, [obj_1, obj_1])
+        assert hash_1 == hash_2
+
+        # Making sure memoization and object identity does not interfere with
+        # hashing
+        # XXX: this fails on joblib master -- not sure what joblib should do in
+        # this case
+        # hash_1, hash_2 = e.map(hash, [[obj_1, obj_1], [obj_1, obj_2]])
+        # assert hash_1 == hash_2
```

Comment:
Would the following assertion pass?:

```python
assert hash([obj_1, obj_1]) == hash([obj_1, obj_2])
```

maybe we have another memoization bug related to string interning.

Reply:
Actually not, it's probably just memoization of the top level object. I think this is kind of expected, no? I am not sure we should do anything about it.

Reply:
> I think this is kind of expected, no? 

My guding principle about hashing is that only value and not identity should play a role in hash values - so personally I would call this behavior a bug. Not sure we should fix this in this PR though.

Reply:
The lists objects are not the same because one has 2 refs to the same object and this is not the case for the other. If we had nested graph structures with cyclic references, e.g. to represent data graphs, having 2 different hashes when only the reference structure changes, this would make sense to me.